src/js/saspowerschoolff.js

```diff
@@ -188,6 +203,85 @@ function login_page () {
     });
 }
 
+function calculate_cumulative_gpa (current_courses) {
+    const list_of_gpas = [];
+    const all_courses = [];
+    const credit_hour_list = [];
+    let element_list = [];
+    const fetches = [];
+    // Fetches grade history page
+    const gpas = fetch("https://powerschool.sas.edu.sg/guardian/termgrades.html")
+        .then(response => response.text())
+        .then(data => {
+            const el = document.createElement("html");
+            el.innerHTML = data;
+            const tabs = el.getElementsByClassName("tabs")[0].getElementsByTagName("li");
+            // Iterate until the end of tabs or until no longer at a high school semester.
+            for (let i = 0; i < tabs.length && /HS$/.test(tabs[i].innerText); i++) {
+                fetches.push(
+                    fetch(tabs[i].getElementsByTagName("a")[0].href)
+                        .then(res => res.text())
+                        .then(function (data2) {
+                            let courses = [];
+                            const semester = document.createElement("html");
+                            semester.innerHTML = data2;
+                            element_list = semester.getElementsByClassName("box-round")[0].getElementsByTagName("table")[0];
+                            element_list = element_list.getElementsByTagName("tbody")[0].getElementsByTagName("tr");
+                            for (let t = 2; t < element_list.length; t++) {
+                                if (element_list[t].innerText.trim() === ("S2")) {
+                                    all_courses.push(courses);
+                                    courses = [];
+                                }
+                                if (element_list[t].getElementsByTagName("th").length > 0) {
+                                    continue;
+                                } else {
+                                    const $prev_course = element_list[t];
+                                    // Creates course object with each course from grade history page
+                                    const course = new Course($prev_course.getElementsByTagName("td")[0].textContent.trim(), "",
+                                        $prev_course.getElementsByTagName("td")[1].textContent.trim(), 0, "");
+
+                                    courses.push(course);
+                                }
+                            }
+                            all_courses.push(courses);
+                        }));
+            }
+            // Calculates cumulative GPA based on credit hours per semester and gpa for each semester.
+            let include_current_semester = false;
+            if (current_courses.length !== 0) {
+                for (let i = 0; i < current_courses.length; i++) {
+                    if (current_courses[i].link.includes("begdate")) {
```

Comment:
The better way to do this would be something like
```suggestion
                    if (new URL(current_courses[i].link).searchParams.get("begdate"))  {
```
but we can deal with that during the cleanup if you'd like.

Reply:
Sure, thanks okay if I merge it now?

Reply:
If the PR's approved by someone with write access, you don't need to ask :)